### PR TITLE
Reduce redundant code common to `Study` and `InTrialStudy`.

### PR DESCRIPTION
**Pull request by sile**
This PR makes `BaseStudy.{best_trial, direction, trials}` methods concrete, and removes redundant code in `Study` and `InTrialStudy`.

### CODE REVIEW

File: optuna/study.py [hunk 1]
```diff
@@ -35,12 +33,17 @@
     ObjectiveFuncType = Callable[[trial_module.Trial], float]
```
Comment (toshihikoyanase) [inline optuna/study.py]: Could you remove the 31st line (`from optuna.distributions import BaseDistribution  # NOQA`), please?
I found that `BaseDistribution` is not used in `study.py` during the review although it seems to be unrelated to this PR directly.

File: optuna/study.py [hunk 2]
```diff
@@ -51,7 +54,7 @@ def best_params(self):
     @property
     def best_value(self):
```
Comment (toshihikoyanase) [inline optuna/study.py]: I found that the 65th line (`raise ValueError('No trials are completed yet.')`) was not covered by the unit tests.
Could you add a test case for it, please?

File: optuna/study.py (hunk 1 above)
Comment (sile) [reply optuna/study.py]: Thank you for your careful reviewing.
Removed: https://github.com/pfnet/optuna/pull/470/commits/16cc5fe3d90bce13544d2a6b156e30b2ad838ef5

File: optuna/study.py (hunk 2 above)
Comment (sile) [reply optuna/study.py]: [NOTE]
We decided to not add a test case for this exception in this PR.
It's not so easy because the public APIs of Optuna doesn't allow creating a trial that has `COMPLETE` state and `None` value at the same time.
